src/autorest.bicep/src/resources.ts

```diff
@@ -0,0 +1,409 @@
+import { keys, Dictionary, values } from 'lodash';
+
+export enum ScopeType {
+  Unknown = 0,
+  Tenant = 1 << 0,
+  ManagementGroup = 1 << 1,
+  Subscription = 1 << 2,
+  ResourceGroup = 1 << 3,
+  Extension = 1 << 4,
+}
+
+export interface ResourceDescriptor {
+  scopeType: ScopeType;
+  namespace: string;
+  typeSegments: string[];
+  apiVersion: string;
+  constantName?: string;
+}
+
+export interface ProviderDefinition {
+  namespace: string;
+  apiVersion: string;
+  resourcesByType: Dictionary<ResourceDefinition[]>;
+}
+
+export interface ResourceDefinition {
+  descriptor: ResourceDescriptor;
+  putRequest: HttpRequest;
+  putParameters: Parameter[];
+  putSchema?: ObjectSchema;
+  getSchema?: ObjectSchema;
+}
+
+const parentScopePrefix = /^.*\/providers\//ig;
+const managementGroupPrefix = /^\/providers\/Microsoft.Management\/managementGroups\/{\w+}\/$/i;
+const tenantPrefix = /^\/$/i;
+const subscriptionPrefix = /^\/subscriptions\/{\w+}\/$/i;
+const resourceGroupPrefix = /^\/subscriptions\/{\w+}\/resourceGroups\/{\w+}\/$/i;
+const resourceGroupMethod = /^\/subscriptions\/{\w+}\/resourceGroups\/{\w+}$/i;
+
+function trimScope(scope: string) {
+  return scope.replace(/\/*$/, '').replace(/^\/*/, '');
+}
+
+function isPathVariable(pathSegment: string) {
+  return pathSegment.startsWith('{') && pathSegment.endsWith('}');
+}
+
+function trimParamBraces(pathSegment: string) {
+  return pathSegment.substr(1, pathSegment.length - 2);
+}
+
+export function getFullyQualifiedType(descriptor: ResourceDescriptor) {
+  return [descriptor.namespace, ...descriptor.typeSegments].join('/');
+}
+
+export function isRootType(descriptor: ResourceDescriptor) {
+  return descriptor.typeSegments.length === 1;
+}
+
+function getHttpRequests(requests: Request[] | undefined) {
+  return requests?.map(x => x.protocol.http as HttpRequest).filter(x => !!x) ?? [];
+}
+
+function hasStatusCode(response: Response, statusCode: string) {
+  const statusCodes = (response.protocol.http as HttpResponse)?.statusCodes;
+  if (!statusCodes) {
+    return;
+  }
+
+  return (statusCodes as string[]).includes(statusCode);
+}
+
+function getNormalizedMethodPath(path: string) {
+  if (resourceGroupMethod.test(path)) {
+    // resource groups are a special case - the swagger API is not defined as a provider API, but they are still deployable in a template as if it was.
+    return "/subscriptions/{subscriptionId}/providers/Microsoft.Resources/resourceGroups/{resourceGroupName}";
+  }
+
+  return path;
+}
+
+export function parseNameSchema<T>(descriptor: ResourceDescriptor, request: HttpRequest, parameters: Parameter[], parseType: (schema: Schema) => T, createConstantName: (name: string) => T) {
+  const path = getNormalizedMethodPath(request.path);
+
+  const finalProvidersMatch = path.match(parentScopePrefix)?.last;
+  if (!finalProvidersMatch) {
+    return { success: false, failureReason: `Unable to locate "/providers/" segment` };
```

Comment:
Might be worth extracting two functions`failure(failureReason: string)` and `success(name: string)`?